### PR TITLE
Ouch again: Go/CouchDB and GopherJS/PouchDB pass (amended) tests

### PR DESCRIPTION
**Pull request by elliott5**
`gopherjs test` failure output below, an edge-case of some kind still to be tracked-down.
```
--- FAIL: TestOuchAll (4.41s)
    --- FAIL: TestOuchAll/qs (2.24s)
        --- FAIL: TestOuchAll/qs/delete_quad (0.19s)
        Error Trace:          tion:0:
        Error:          Not equal:
                        expected: []quad.Quad(nil)
                        actual: []quad.Quad{}

                        Diff:
                        --- Expected
                        +++ Actual
                        @@ -1,2 +1,3 @@
                        -([]quad.Quad) <nil>
                        +([]quad.Quad) {
                        +}
```

Please note: in values.go:toOuchDoc() I had to create fields from the nosql nodeValue type since
the fields are used by the test code compare_typed_values, but the CouchDB $ne comparator does not work correctly for missing fields. A more elegant way of doing this may be required!

See README & comments for other things still to do :)

---
This change is [Reviewable](https://reviewable.io/reviews/dennwc/cayley/4)

### CODE REVIEW

**Comment by dennwc** [comment]
Made a quick review pass. But to get all specifics I'll need to run tests and see how it behaves currently. So I will make a second pass later.

---

Reviewed 6 of 14 files at r3.
Review status: 6 of 8 files reviewed at latest revision, 17 unresolved discussions.

---

*[graph/nosql/ouch/ouch.go, line 230 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-frORzJcgKmr2Us9pT:-L-frORzJcgKmr2Us9pU:bdfkio4) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L230)):*
> ```Go
> 		ouchQuery: map[string]interface{}{
> 			"selector": map[string]interface{}{},
> 			"limit":    1000000, // million row limit, default is 25
> ```

Hmm, interesting. It always limits a number of results to 25 if limit is not set?

---

*[graph/nosql/ouch/ouch.go, line 301 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fmNabRwK_-Nqayz5g:-L-fmNabRwK_-Nqayz5h:bud1e3b) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L301)):*
> ```Go
> 				test = "$lte"
> 			default:
> 				msg := "unknown nosqlFilter " + fmt.Sprintf("%v", filter.Filter)
> ```

you may place this text inside `Sprintf` as well

---

*[graph/nosql/ouch/ouch.go, line 323 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fmZmBo8QwGdBA87dX:-L-fmZmBo8QwGdBA87dY:b-3dbwae) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L323)):*
> ```Go
> func (q *Query) Count(ctx context.Context) (int64, error) {
> 	dpanic("Query.Count")
> 	it := q.Iterate()
> ```

it's better to type-assert it to `*Iterator` here, so you can omit it below

---

*[graph/nosql/ouch/ouch.go, line 417 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-frhl72nL3E2BRfj4D:-L-frhl72nL3E2BRfj4E:bb7be76) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L417)):*
> ```Go
> 	var id string
> 	var haveID bool
> 	id, haveID = it.doc[idField].(string)
> ```

`:=`

---

*[graph/nosql/ouch/ouch.go, line 434 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-frw3y0KWEk4WCNQ0f:-L-frw3y0KWEk4WCNQ0g:bdwnylr) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L434)):*
> ```Go
> 	}
> 	if !it.hadNext {
> 		it.err = errors.New("Iterator.Doc called before Iterator.Next")
> ```

it's okay to just return `nil` here as in if branch above

---

*[graph/nosql/ouch/ouch.go, line 457 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fn9ZabMweC-2YNuQ6:-L-fn9ZabMweC-2YNuQ7:b-k6vtsd) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L457)):*
> ```Go
> 	db   *DB
> 	col  string
> 	q    nosql.Query
> ```

Can be changed to `*Query`

---

*[graph/nosql/ouch/ouch.go, line 478 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fsNUtD6bWqQzvl8Ki:-L-fsNUtD6bWqQzvl8Kj:b-6q7af5) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L478)):*
> ```Go
> 
> 	if len(d.keys) > 0 {
> 		d.q.(*Query).ouchQuery["selector"].(map[string]interface{})[idField] = map[string]interface{}{"$in": d.keys}
> ```

for `len(d.keys) == 1` is can be set to equality instead of `$in`

---

*[graph/nosql/ouch/ouch.go, line 491 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fsdicDwd0SXlUJrxh:-L-fsdicDwd0SXlUJrxi:b-d1peyl) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L491)):*
> ```Go
> 
> 	for it.Next(ctx) {
> 		if it.err != nil {
> ```

it's okay to check if once after the loop, since in case of error `Next` will cancel the loop immediately and this check will have no effect

---

*[graph/nosql/ouch/ouch.go, line 548 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-ft--iVBOaBSZh6DmK:-L-ft--jk7Ec6lUnK7Dk:b-bqc7ex) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch.go#L548)):*
> ```Go
> 		if !u.upsert {
> 			return err
> 		} else {
> ```

omit else branch - if branch returns anyway

---

*[graph/nosql/ouch/ouch_test.go, line 21 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-ftqc83pLZvC3wUPUl:-L-ftqc83pLZvC3wUPUm:bqz656a) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch_test.go#L21)):*
> ```Go
> )
> 
> const remote = "http://testusername:testpassword@127.0.0.1:5984/ouchtest"
> ```

I could make it run under Docker automatically while trying it locally. Do you have Docker installed, or should I leave local test option as well?

---

*[graph/nosql/ouch/ouch_test.go, line 46 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fnR62GGPPqAMzQoos:-L-fnR62GGPPqAMzQoot:b-jb09jw) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch_test.go#L46)):*
> ```Go
> 			return os.RemoveAll(testDBname)
> 		} else {
> 			client, err := kivik.New(context.TODO(), defaultDriverName, testDBname)
> ```

move ctx to the top, and use it in all function calls: `ctx := context.TODO()`

---

*[graph/nosql/ouch/ouch_test.go, line 98 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fuPo5XFIQr5eqEIv5:-L-fuPo5XFIQr5eqEIv6:b-n0vkhs) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch_test.go#L98)):*
> ```Go
> var allsorts = nosql.Document{
> 	"Vnil": nil,
> 	"VDocument": nosql.Document{
> ```

I see now that I forgot to add tests for inserting nested fields and binary fields

---

*[graph/nosql/ouch/ouch_test.go, line 130 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fw1Sr4zVxWxwKuI6h:-L-fw1Sr4zVxWxwKuI6i:bbni5y6) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch_test.go#L130)):*
> ```Go
> 		}
> 
> 		dbc, err := Open(dbName, graph.Options{})
> ```

passing nil Options should be safe as well

---

*[graph/nosql/ouch/values.go, line 43 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fkOz_0_XkPHds6sAi:-L-fkOz_0_XkPHds6sAj:b-s7up22) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/values.go#L43)):*
> ```Go
> 	case nil:
> 		return nil
> 	case nosql.Strings: // special handling here, as type can't be inferred from json
> ```

This was actually the reason why I dropped `nosql.Key` for `Value` interface - it's now the only array type that is supported. So if kivik can store json array, this can be an indication that `Strings` is stored there.

---

*[graph/nosql/ouch/values.go, line 68 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-fkqWTiMpKg7rAOAM7:-L-fkqWUmDhD9z6Pq_bs:b-elxk4c) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/values.go#L68)):*
> ```Go
> 		return nil
> 	}
> 	m := map[string]interface{}{}
> ```

`make(map[string]interface{})`

---

*[graph/nosql/ouch/values.go, line 80 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-flc0EbSBwXxeEcV4e:-L-flc0EbSBwXxeEcV4f:b6loljm) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/values.go#L80)):*
> ```Go
> 
> 	for k, v := range d {
> 		if len(k) > 0 {
> ```

invert condition and lower the code:
```golang
if len(k) == 0 {
  continue
}
// ... rest of the code
```

---

*[graph/nosql/ouch/values.go, line 210 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-flJSf3a3vVEgrtEuE:-L-flJSf3a3vVEgrtEuF:b-yz64us) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/values.go#L210)):*
> ```Go
> 	m := make(nosql.Document, len(d))
> 	for k, v := range d {
> 		switch k {
> ```

```golang
switch k {
case "", idField, revField, collectionField:
  continue
}
// ... rest of the code
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4#-:-L-fqHhon_l6pygCJEEG:b6crqf0)*

**Comment by elliott5** [comment]
Yes I have Docker installed, so that's the better route for testing I'm sure ... I just don't know how to do it :)

**Comment by dennwc** [comment]
Reviewed 2 of 2 files at r4.
Review status: 7 of 9 files reviewed at latest revision, 18 unresolved discussions.

---

*[graph/graphtest/graphtest.go, line 142 at r4](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-g5qZ-iZKEC6XJNiHP:-L-g5qZ-iZKEC6XJNiHQ:bolu76x) ([raw file](https://github.com/dennwc/cayley/blob/10d964e5f3a9ef26d9a758cee511ae69d9b24db1/graph/graphtest/graphtest.go#L142)):*
> ```Go
> 	sort.Sort(res)
> 	if res == nil {
> 		return []quad.Quad(nil) // GopherJS seems to have a bug with this type conversion for a nil value
> ```

it doesn't seems related to Gopher JS or this particular change
`res` is a local variable and only this function can modify it... and only place can - it's `append` above
in case `append` was not called, res will be nil and will be returned as nil; if append was done, it will not be nil
and `Sort` cannot switch value from nil as well

so if you see an error after this call it may be an empty quad inserted to slice, for example... I don't see any other possibility
or maybe an outdated version of `require` might be the case - try running `go get -u github.com/stretchr/testify/require`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4)*

**Comment by elliott5** [comment]
To understand the issue, you need to look at the generated JS (which seems wrong to me):
```
	if res == nil {
		return []quad.Quad(nil) // GopherJS seems to have a bug with this type conversion for a nil value
	}
	return res
```
generates JS code:
```
	if (res === quad.ByQuadString.nil) {
		$s = -1; return (sliceType$5.nil);
	}
	$s = -1; return $subslice(new sliceType$5(res.$array), res.$offset, res.$offset + res.$length);
```
and $subslice does not deal with nil:
```
var $subslice = function(slice, low, high, max) {
  if (low < 0 || high < low || max < high || high > slice.$capacity || max > slice.$capacity) {
    $throwRuntimeError("slice bounds out of range");
  }
  var s = new slice.constructor(slice.$array);
  s.$offset = slice.$offset + low;
  s.$length = slice.$length - low;
  s.$capacity = slice.$capacity - low;
  if (high !== undefined) {
    s.$length = high - low;
  }
  if (max !== undefined) {
    s.$capacity = max - low;
  }
  return s;
};
```
So I need to make an error report to the GopherJS team to add-in the nil test.

**Comment by dennwc** [comment]
Review status: 7 of 9 files reviewed at latest revision, 17 unresolved discussions.

---

*[graph/graphtest/graphtest.go, line 142 at r4](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-g5qZ-iZKEC6XJNiHP:-L-g_hu7WuGl3Qg51tvQ:bnmkaq6) ([raw file](https://github.com/dennwc/cayley/blob/10d964e5f3a9ef26d9a758cee511ae69d9b24db1/graph/graphtest/graphtest.go#L142)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

it doesn't seems related to Gopher JS or this particular change
`res` is a local variable and only this function can modify it... and only place can - it's `append` above
in case `append` was not called, res will be nil and will be returned as nil; if append was done, it will not be nil
and `Sort` cannot switch value from nil as well

so if you see an error after this call it may be an empty quad inserted to slice, for example... I don't see any other possibility
or maybe an outdated version of `require` might be the case - try running `go get -u github.com/stretchr/testify/require`
</blockquote></details>

Makes total sense now :) Was assuming GopherJS will just pass the same variable to return directly.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4)*

**Comment by dennwc** [comment]
Reviewed 2 of 4 files at r5.
Review status: 7 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[graph/nosql/ouch/values.go, line 210 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-flJSf3a3vVEgrtEuE:-L-gpIdRAeo4dDXwSUYo:brab05w) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/values.go#L210)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

```golang
switch k {
case "", idField, revField, collectionField:
  continue
}
// ... rest of the code
```
</blockquote></details>

`if len(k) > 0 {` is not necessary anymore - it's covered by `case ""`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4)*

**Comment by elliott5** [comment]
FYI see https://github.com/go-kivik/pouchdb/issues/19 on the size of the generated JS code via kivik

**Comment by elliott5** [comment]
FYI, the ouch GopherJS/PouchDB driver for Cayley adds around 1.4Mb more minimized JS than just using a memory graph. 

1.3Mb of this increase relates to the use of the protobuf protocol by Cayley, where  graph/nosql/quadstore.go uses quad/pquad to encode/decode quad values in a switch. It is too complex to remove this immediately, during the nosql creation phase, a future optimisation perhaps. 

This JS size issue may limit the practical applications for this code until the WASM compiler back-end is complete and a further rewrite can reduce the download size.

**Comment by elliott5** [comment]
With the protobuf JS size issue (described above) and the advent of Christmas I'm parking my work on this until next year.

But still happy to rework the test environment further to remove my bootstrap tests if required before you accept the PR.

**Comment by elliott5** [comment]
I see your nosql branch has changed, I'll bring this PR back in line tomorrow.

**Comment by dennwc** [comment]
Thanks @elliott5 ! I will also push context change later today + more tests for nosql.

---

Reviewed 1 of 3 files at r6, 1 of 2 files at r7, 1 of 4 files at r8, 3 of 4 files at r9, 1 of 2 files at r10, 2 of 16 files at r11.
Review status: 6 of 20 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[graph/nosql/ouch/ouch_test.go, line 56 at r6](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-l6b7kWy83xPZ-FaY9:-L-l6b7kWy83xPZ-FaYA:bgtek3e) ([raw file](https://github.com/dennwc/cayley/blob/1d13d36c52ec631109aa916fa3aae77210449223/graph/nosql/ouch/ouch_test.go#L56)):*
> ```Go
> 					dbDir := thisFile.Name() // TODO Windows tests
> 					//fmt.Println("DEBUG remove all:", dbDir)
> 					if err := os.RemoveAll(dbDir); err != nil {
> ```

We usually do it this way:
Suppose you need a `DB` interface for test. You can define constructor like this:
```golang
func NewDB(t testing.TB) (DB, func()) {
   dir, err := ioutil.TmpDir("", "pouch-")
   if err != nil {
     t.Fatal(err)
   }
   return SomeDB{dir:dir}, func(){
     os.RemoveAll(dir)
   }
}
```
First, note that error handling is done in the constructor itself and applied to testing object directly, so caller will not have to check errors all the time.
Second, note the function that is returned as a second argument - it's a cleanup function that can remove documents, remove tmp files, close DB, remove docker container, etc.

And the test will look like this:
```golang
func TestFoo(t *testing.T) {
   db, closer := NewDB(t)
   defer closer()

   // ... test stuff
}
```

And if you have multiple implementations, you can generalize the test and call it with different constructors:
```golang
type DBFunc func(t testing.TB) (DB, func())

// generic test
func TestDB(t *testing.T, fnc DBFunc) {
   db, closer := fnc(t)
   defer closer()

   // ... test stuff
}

// implementation tests
func TestPouch(t *testing.T) {
   TestDB(t, NewPouch)
}
func TestCouch(t *testing.T) {
   TestDB(t, NewCouch)
}
```
And each implementation may change how cleanup function works, so tests won't need to care about cleanup.

---

*[graph/nosql/ouch/ouch_test.go, line 97 at r6](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-l9Hnxr1QZzZI9hlwS:-L-l9Hnxr1QZzZI9hlwT:bqpnahr) ([raw file](https://github.com/dennwc/cayley/blob/1d13d36c52ec631109aa916fa3aae77210449223/graph/nosql/ouch/ouch_test.go#L97)):*
> ```Go
> 	}
> 
> 	// TODO review: this compaction phase is fun, but adds needless complexity & is not run for most tests
> ```

Is it for speed only?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4#-:-L0WBEDZ4Y5Psc1jUCwO:bn4mb3s)*

**Comment by dennwc** [comment]
Review status: 6 of 20 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/dennwc/cayley/4#-L0XpgeoB4Z1i4rGIoBq:-L0XpgeoB4Z1i4rGIoBr:b-4b1adb) (no related file):*
It's now safe to rebase onto `dennwc/nosql`. I expect no hard changes to APIs.

Please make sure to squash all unrelated commits - I see nearly half of them twice in the PR. I can assist with this process via Slack.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4)*

**Comment by elliott5** [comment]
Review status: 6 of 20 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[a discussion](https://reviewable.io:443/reviews/dennwc/cayley/4#-L0XpgeoB4Z1i4rGIoBq:-L0Y9whTAXaTYLTTrrCu:bmor4jj) (no related file):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

It's now safe to rebase onto `dennwc/nosql`. I expect no hard changes to APIs.

Please make sure to squash all unrelated commits - I see nearly half of them twice in the PR. I can assist with this process via Slack.
</blockquote></details>

A slack discussion would of course be wonderful, meantime I'll battle on.

---

*[graph/nosql/ouch/ouch_test.go, line 21 at r3](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-ftqc83pLZvC3wUPUl:-L0YPZid3dK24bBx6xxw:buoq982) ([raw file](https://github.com/dennwc/cayley/blob/32b140ecc09569305ad8ba51ebc729743786a461/graph/nosql/ouch/ouch_test.go#L21)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

I could make it run under Docker automatically while trying it locally. Do you have Docker installed, or should I leave local test option as well?
</blockquote></details>

I have docker installed, but GopherJS/PouchDB requires running locally.

---

*[graph/nosql/ouch/ouch_test.go, line 56 at r6](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-l6b7kWy83xPZ-FaY9:-L0YPAYPpEeBPlB1SMrk:b-2i7k6p) ([raw file](https://github.com/dennwc/cayley/blob/1d13d36c52ec631109aa916fa3aae77210449223/graph/nosql/ouch/ouch_test.go#L56)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

We usually do it this way:
Suppose you need a `DB` interface for test. You can define constructor like this:
```golang
func NewDB(t testing.TB) (DB, func()) {
   dir, err := ioutil.TmpDir("", "pouch-")
   if err != nil {
     t.Fatal(err)
   }
   return SomeDB{dir:dir}, func(){
     os.RemoveAll(dir)
   }
}
```
First, note that error handling is done in the constructor itself and applied to testing object directly, so caller will not have to check errors all the time.
Second, note the function that is returned as a second argument - it's a cleanup function that can remove documents, remove tmp files, close DB, remove docker container, etc.

And the test will look like this:
```golang
func TestFoo(t *testing.T) {
   db, closer := NewDB(t)
   defer closer()

   // ... test stuff
}
```

And if you have multiple implementations, you can generalize the test and call it with different constructors:
```golang
type DBFunc func(t testing.TB) (DB, func())

// generic test
func TestDB(t *testing.T, fnc DBFunc) {
   db, closer := fnc(t)
   defer closer()

   // ... test stuff
}

// implementation tests
func TestPouch(t *testing.T) {
   TestDB(t, NewPouch)
}
func TestCouch(t *testing.T) {
   TestDB(t, NewCouch)
}
```
And each implementation may change how cleanup function works, so tests won't need to care about cleanup.
</blockquote></details>

Excellent! I'll use this approach.

---

*[graph/nosql/ouch/ouch_test.go, line 97 at r6](https://reviewable.io:443/reviews/dennwc/cayley/4#-L-l9Hnxr1QZzZI9hlwS:-L0YOZYJesXfPI1pQdpc:b-8jbijl) ([raw file](https://github.com/dennwc/cayley/blob/1d13d36c52ec631109aa916fa3aae77210449223/graph/nosql/ouch/ouch_test.go#L97)):*
<details><summary><i>Previously, dennwc (Denys Smirnov) wrote…</i></summary><blockquote>

Is it for speed only?
</blockquote></details>

yes, well space actually in live operation - I'll remove it from the tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4)*

**Comment by elliott5** [comment]
@dennwc Happy New Year! 
Do let me know if you need more from me in order to move the "ouch" driver into the main project.

**Comment by dennwc** [comment]
I've enabled Docker tests for CouchDB in `dennwc/ouch` branch and rebased everything.

JS tests might be broken because of new Docker dependency in tests, but I will split that out as soon as I get JS part running locally.

One feature that is missing now is regexp support that was added in master recently. It seems like CouchDB support regexp natively, so we only need to convert it from Go pattern to one accepted by Couch. In case Pouch has no support for it, we can still add it as a filter on Go side. 

I will close this PR now, since it's a bit broken anyway. Feel free to open PRs against `dennwc/ouch` if there is something you want to modify.

I will also alter Git history of my branch to include correct author (you) for first commit once everything is ready to merge.

---

Review status: 0 of 24 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dennwc/cayley/4#-:-L2kiGAWCs9KTzpZvU6Q:bcppn2w)*

**Comment by elliott5** [comment]
Many thanks @dennwc !

As you say, it looks like mango queries (that this driver generates) accept $regex as selection criteria for both Couch & Pouch, so it is only a matter of the difference between Go and Erlang syntax for them (if any) -  http://docs.couchdb.org/en/2.0.0/api/database/find.html#condition-operators ... then working through implementation differences between the two DBs (if any).

Let me know if you need any help, otherwise I sense you will be faster doing the work alone.
